### PR TITLE
[Delivers #168673139] Migrate Predictions category to use aws-sdk v3 clients

### PR DESCRIPTION
**Pull request by jordanranz**
Related to #3365

_Description of changes:_
This PR upgrades the aws-sdk v2 clients used in the Predictions category to aws-sdk V3 client. This helps achieve modularization at the AWS clients level in Predictions.

[Delivers #168673139]

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/predictions/src/Providers/AmazonAIIdentifyPredictionsProvider.ts [hunk 1]
```diff
@@ -256,17 +288,23 @@ export class AmazonAIIdentifyPredictionsProvider extends AbstractIdentifyPredict
 	 * @return {Promise<IdentifyLabelsOutput>} - Promise resolving to organized detectModerationLabels response.
 	 */
 	private detectModerationLabels(
-		param: Rekognition.DetectFacesRequest
+		param: DetectModerationLabelsInput
```
Comment (elorzafe) [inline packages/predictions/src/Providers/AmazonAIIdentifyPredictionsProvider.ts]: Good catch!

File: packages/predictions/src/Providers/AmazonAIInterpretPredictionsProvider.ts [hunk 2]
```diff
@@ -139,7 +150,8 @@ export class AmazonAIInterpretPredictionsProvider extends AbstractInterpretPredi
 
 	private detectKeyPhrases(params, comprehend): Promise<Array<KeyPhrases>> {
```
Comment (elorzafe) [inline packages/predictions/src/Providers/AmazonAIInterpretPredictionsProvider.ts]: I like the refactor to `comprehendClient` on `interpretText` can you do here as well? :)

File: packages/predictions/src/types/Predictions.ts [hunk 3]
```diff
@@ -352,3 +352,187 @@ export function isInterpretTextInput(obj: any): obj is InterpretTextInput {
...
+	Bytes?: ArrayBuffer | ArrayBufferView | string;
+
+	/**
+	 * <p>Identifies an S3 object as the image source.</p>
+	 */
+	S3Object?: S3Object;
+}
+
+export type ImageBlob = Buffer | Uint8Array | Blob | string;
+export type S3Bucket = string;
```
Comment (elorzafe) [inline packages/predictions/src/types/Predictions.ts]: I wouldn't export this type are specific to the service and this should be agnostic.

File: packages/predictions/src/types/Predictions.ts [hunk 4]
```diff
@@ -352,3 +352,187 @@ export function isInterpretTextInput(obj: any): obj is InterpretTextInput {
...
+export interface S3Object {
```
Comment (elorzafe) [inline packages/predictions/src/types/Predictions.ts]: ditto

File: packages/predictions/src/types/Predictions.ts [hunk 5]
```diff
@@ -352,3 +352,187 @@ export function isInterpretTextInput(obj: any): obj is InterpretTextInput {
...
+	/**
+	 * S3 object key name.
+	 */
+	Name?: S3ObjectName;
+	/**
+	 * If the bucket is versioning enabled, you can specify the object version. 
+	 */
+	Version?: S3ObjectVersion;
+}
+export type S3ObjectName = string;
```
Comment (elorzafe) [inline packages/predictions/src/types/Predictions.ts]: ditto

File: packages/predictions/src/types/Predictions.ts [hunk 6]
```diff
@@ -352,3 +352,187 @@ export function isInterpretTextInput(obj: any): obj is InterpretTextInput {
...
+export type S3ObjectVersion = string;
```
Comment (elorzafe) [inline packages/predictions/src/types/Predictions.ts]: ditto

File: packages/predictions/src/types/Predictions.ts [hunk 7]
```diff
@@ -352,3 +352,187 @@ export function isInterpretTextInput(obj: any): obj is InterpretTextInput {
...
+	| "PAGE"
+	| "LINE"
+	| "WORD"
+	| "TABLE"
+	| "CELL"
+	| "SELECTION_ELEMENT"
+	| string;
+
+	/**
+	 * <p>The confidence that Amazon Textract has in the accuracy of the recognized text and the accuracy of the geometry points around the recognized text.</p>
```
Comment (elorzafe) [inline packages/predictions/src/types/Predictions.ts]: I would remove the piece that says Amazon or AWS from this comments

File: packages/predictions/src/types/Predictions.ts [hunk 8]
```diff
@@ -352,3 +352,187 @@ export function isInterpretTextInput(obj: any): obj is InterpretTextInput {
...
+	 */
+	BoundingBox?: BoundingBox;
+
+	/**
+	 * <p>Within the bounding box, a fine-grained polygon around the detected text.</p>
+	 */
+	Polygon?: Array<Point> | Iterable<Point>;
+}
+export type BlockList = Block[];
+export interface Block {
```
Comment (elorzafe) [inline packages/predictions/src/types/Predictions.ts]: I think the types related to Amazon or AWS should be under Provider no on Category types.

**Comment by iartemiev** [comment]
Tests are running fine locally but getting this error in CircleCI:

```
lerna ERR! yarn run test stderr:
FAIL __tests__/Providers/AWSKinesisProvider-unit-test.ts
  ● Test suite failed to run

    ENOMEM: not enough memory, read
```

**Comment by iartemiev** [comment]
Looks like the memory issue was just a fluke. I only removed a period from a comment in the last commit.

If we run into it again though, we can try: https://github.com/zcorky/moment/commit/dd9ebd863aa94e7e26f5864bd8005c9eff06b8e6

File: packages/predictions/__tests__/Providers/AWSAIConvertPredictionsProvider-unit-test.ts [hunk 9]
```diff
@@ -1,46 +1,33 @@
...
-import * as Translate from 'aws-sdk/clients/translate';
-import * as TextToSpeech from 'aws-sdk/clients/polly';
-
-jest.mock('aws-sdk/clients/translate', () => {
-	const Translate = () => {
-		return;
-	};
-	const result = { TranslatedText: 'translatedText', TargetLanguageCode: 'es' };
-	Translate.prototype.translateText = (params, callback) => {
+import { TranslateClient } from '@aws-sdk/client-translate-browser/TranslateClient';
```
[the remark this answers is not in the document]
Comment (jordanranz) [reply packages/predictions/__tests__/Providers/AWSAIConvertPredictionsProvider-unit-test.ts]: I believe the Signing process for v3 migration will be handled in: https://github.com/aws-amplify/amplify-js/pull/4077. I don't think there is a difference between the actual client implementation between v2 and v3 that would prevent it from working in React Native. It is a valid concern that should definitely be tested.

File: packages/predictions/__tests__/Providers/AWSAIIdentifyPredictionsProvider-unit-test.ts [hunk 10]
```diff
@@ -556,16 +571,17 @@ describe('Predictions identify provider test', () => {
...
 		test('happy case input source valid public s3object', () => {
 			const detectLabelInput: IdentifyLabelsInput = {
 				labels: { source: { level: 'public', key: 'key' }, type: 'LABELS' },
 			};
 			jest
-				.spyOn(Rekognition.prototype, 'detectLabels')
+				.spyOn(RekognitionClient.prototype, 'send')
 				.mockImplementationOnce((input, callback) => {
-					expect(input.Image.S3Object.Name).toMatch('public/key');
+					// expect(input.Image.S3Object.Name).toMatch('public/key');
```
[the remark this answers is not in the document]
Comment (jordanranz) [reply packages/predictions/__tests__/Providers/AWSAIIdentifyPredictionsProvider-unit-test.ts]: Ahh good catch, I can fix these. `input` is now part of the `command` passed in.

File: packages/predictions/src/Providers/AmazonAIInterpretPredictionsProvider.ts (hunk 2 above)
Comment (jordanranz) [reply packages/predictions/src/Providers/AmazonAIInterpretPredictionsProvider.ts]: Which refactor on interpretText?

File: packages/predictions/src/Providers/AmazonAIInterpretPredictionsProvider.ts [hunk 11]
```diff
@@ -139,7 +150,8 @@ export class AmazonAIInterpretPredictionsProvider extends AbstractInterpretPredi
...
 		return new Promise((res, rej) => {
-			comprehend.detectKeyPhrases(params, (err, data) => {
+			const detectKeyPhrasesCommand = new DetectKeyPhrasesCommand(params);
+			comprehend.send(detectKeyPhrasesCommand, (err, data) => {
```
Comment (pravgupt) [inline packages/predictions/src/Providers/AmazonAIInterpretPredictionsProvider.ts]: Same, this is not changed to async/awaits

File: packages/predictions/src/types/Providers/AmazonAIIdentifyPredictionsProvider.ts [hunk 12]
```diff
@@ -0,0 +1,185 @@
+import { BoundingBox, Point } from '../Predictions';
```
Comment (pravgupt) [inline packages/predictions/src/types/Providers/AmazonAIIdentifyPredictionsProvider.ts]: I didn't get this file's purpose or the name. Any reason why these types cannot be part of the Predections.ts in the types directory?

Comment (jordanranz) [reply packages/predictions/src/types/Providers/AmazonAIIdentifyPredictionsProvider.ts]: These are type that are AWS model specific. See the @elorzafe's comments above.

File: packages/predictions/src/Providers/AmazonAIIdentifyPredictionsProvider.ts [hunk 13]
```diff
@@ -222,14 +258,15 @@ export class AmazonAIIdentifyPredictionsProvider extends AbstractIdentifyPredict
...
 	 * @return {Promise<IdentifyLabelsOutput>} - Promise resolving to organized detectLabels response.
 	 */
 	private detectLabels(
-		param: Rekognition.DetectLabelsRequest
+		param: DetectLabelsInput
 	): Promise<IdentifyLabelsOutput> {
 		return new Promise((res, rej) => {
-			this.rekognition.detectLabels(param, (err, data) => {
+			const detectLabelsCommand = new DetectLabelsCommand(param);
+			this.rekognitionClient.send(detectLabelsCommand, (err, data) => {
```
[the remark this answers is not in the document]
Comment (jordanranz) [reply packages/predictions/src/Providers/AmazonAIIdentifyPredictionsProvider.ts]: Fixing this.

File: packages/predictions/src/Providers/AmazonAIInterpretPredictionsProvider.ts (hunk 11 above)
Comment (jordanranz) [reply packages/predictions/src/Providers/AmazonAIInterpretPredictionsProvider.ts]: Fixing.

File: packages/predictions/src/types/Providers/AmazonAIIdentifyPredictionsProvider.ts (hunk 12 above)
Comment (pravgupt) [reply packages/predictions/src/types/Providers/AmazonAIIdentifyPredictionsProvider.ts]: I think the name is misleading for the file name and the directory. If the types are specific to that provider (i.e. a specific implementation AmazonAIIdentifyPredictionsProvider provider) we should create a local types directory their (src/Providers/types/<this file>) and possibly rename to reflect it's a types file and not a provider implementation.

If it's a generic types file which can/should be used by any Identify provider, then we can keep it here but rename it to reflect this.

Comment (jordanranz) [reply packages/predictions/src/types/Providers/AmazonAIIdentifyPredictionsProvider.ts]: Fair. I can either move this inside of `src/Providers/types/` and make it more clear that they are types for this Provider or I can move the AWS specific types into a common AWS types files. What do you think @elorzafe @pravgupt?

Comment (pravgupt) [reply packages/predictions/src/types/Providers/AmazonAIIdentifyPredictionsProvider.ts]: I prefer the first option as these are not generic types.

Comment (jordanranz) [reply packages/predictions/src/types/Providers/AmazonAIIdentifyPredictionsProvider.ts]: I think they are technically generic throughout AWS. Multiple AWS Services will utilize these types. For example, the `Image` type is used by both Textract and Rekognition. I ended up using an AWSTypes file.

File: packages/predictions/src/Providers/IdentifyTextUtils.ts [hunk 14]
```diff
@@ -22,7 +18,7 @@ function getBoundingBox(geometry: Geometry): BoundingBox {
 
 function getPolygon(geometry: Geometry): Polygon {
 	if (!geometry) return undefined;
-	return makeCamelCaseArray(geometry.Polygon);
+	return makeCamelCaseArray(Array.from(geometry.Polygon));
```
Comment (elorzafe) [inline packages/predictions/src/Providers/IdentifyTextUtils.ts]: Nice!

File: packages/predictions/src/Providers/AmazonAIIdentifyPredictionsProvider.ts [hunk 15]
```diff
@@ -319,131 +320,138 @@ export class AmazonAIIdentifyPredictionsProvider extends AbstractIdentifyPredict
...
+						'Smile',
+						'Eyeglasses',
+						'Sunglasses',
+						'Gender',
+						'Beard',
+						'Mustache',
+						'EyesOpen',
+						'MouthOpen',
+					];
+					const faceAttributes = makeCamelCase(detail, attributeKeys);
```
[the remark this answers is not in the document]
Comment (jordanranz) [reply packages/predictions/src/Providers/AmazonAIIdentifyPredictionsProvider.ts]: https://github.com/aws-amplify/amplify-js/pull/4084/files/efc95b08792013fec0bf57a911e395faf5611763..f57efa9120f30a1780a863397f96f5487b2dbc3f#diff-4b5891b38433c8f57fff1d478d4a751dR435

[the remark this answers is not in the document]
Comment (elorzafe) [reply packages/predictions/src/Providers/AmazonAIIdentifyPredictionsProvider.ts]: but is never readed

**Pull request merged**